### PR TITLE
[FlexAttention] Add BlockMask.adjust to facilitate block_mask reusage

### LINKED ISSUE #136232: NaN in Flex Attention backward if BlockMask is larger than first run seq_len when return_lse=True and torch compiled
### 🐛 Describe the bug

We've came across a NaN in the backward pass of Flex Attention. In order to trigger it, you need all of the following:

- BlockMask slightly larger than actual first run's seq_len
- return_lse = True
- Torch compile

```py
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
torch.set_default_device('cuda')
def causal_mask(b, h, q_idx, kv_idx):
    causal_mask = q_idx >= kv_idx
    return causal_mask
def create_qkv(seq_len):
    q = torch.randn(1, 12, seq_len, 64, requires_grad=True)
    k = torch.randn(1, 12, seq_len, 64, requires_grad=True)
    v = torch.randn(1, 12, seq_len, 64, requires_grad=True)
    return q, k, v
causal_block_mask = create_block_mask(causal_mask, B=None, H=None, Q_LEN=4096, KV_LEN=4096)
def forward(q, k, v, mask):
    out, lse = flex_attention(q, k, v, block_mask=mask, return_lse=True)
    return out * torch.exp(lse)[..., None]
def repro(seq_len, backend):
    print("Repro at seq_len", seq_len)
    compiled = torch.compile(forward, backend=backend)
    for i in range(10):
        q, k, v = create_qkv(seq_len)
        out = compiled(q, k, v, causal_block_mask)
        loss = out.sum()
        loss.backward()
        assert not q.grad.isnan().any(), "Q has NaN"
        assert not k.grad.isnan().any(), "K has NaN"
        assert not v.grad.isnan().any(), "V has NaN"
    print("Done")

for backend in ["inductor", "eager"]:
    print(backend, "4096 then 4096 - 128")
    torch.compiler.reset()
    repro(4096, backend)
    repro(4096 - 128, backend)

    print(backend, "just 4096 - 128")
    torch.compiler.reset()
    try:
        repro(4096 - 128, backend)
    except AssertionError as e:
        print("ERROR", e)
```

```
inductor 4096 then 4096 - 128
Repro at seq_len 4096
Done
Repro at seq_len 3968
Done
inductor just 4096 - 128
Repro at seq_len 3968
ERROR K has NaN
eager 4096 then 4096 - 128
Repro at seq_len 4096
Done
Repro at seq_len 3968
Done
eager just 4096 - 128
Repro at seq_len 3968
Done
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @ydwu4 @bdhirsh @Chillee @drisspg @yanboliang @BoyuanFeng

### Versions

'2.6.0.dev20240916+cu124'

### PR DESCRIPTION
**Pull request by yanboliang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136601

Fixes #136232
The root cause of #136232 is because of ```block_mask``` reusage, which leads a wrong ```full_q_num_blocks``` or ```full_q_indices``` under some ```mask_mod```, e.g, causal mask. Let me use the following example to demonstrate the issue:
```Shell
block_mask_large = create_block_mask(causal_mask, 1, 1, 2048, 2048)
block_mask_small = create_block_mask(causal_mask, 1, 1, 1024, 1024)
block_mask_adjusted = block_mask_large.adjust(1024, 1024)
(Pdb) block_mask_large.full_q_num_blocks
tensor([[[15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0]]],
       device='cuda:0', dtype=torch.int32)
(Pdb) block_mask_large.full_q_num_blocks[:,:,:8]                            <- second way
tensor([[[15, 14, 13, 12, 11, 10,  9,  8]]], device='cuda:0',
       dtype=torch.int32)
(Pdb) block_mask_small.full_q_num_blocks
tensor([[[7, 6, 5, 4, 3, 2, 1, 0]]], device='cuda:0', dtype=torch.int32)
```
Inside of the Triton kernel, we are using the second way to get corresponding ```full_q_num_blocks```, which accesses wrong indices data. I think our assumption of slicing the original large block_mask only valid when handling ```kv_num_blocks``` and ```kv_indices```, the ```q_num_blocks``` and ```q_indices``` are transposed, so we can't naively slice it. The ```kv_***``` only used in backward, which explains why this issue was not disclosed in the previous tests, since most of them are decoding use cases.
We can fix this issue inside of the kernel, but I don't like that idea, since the transpose logics only valid for a specific ```mask_mods```, the kernel and ```mask_mods``` should be orthogonal, and we should not put ```mask_mod``` specific logic inside of the kernel. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by cora-codes** [comment]
Hm, I feel like this might be incomplete, it doesn't fully fix my issue. If I'm understanding right, my script will still NaN in the same scenario (due to reading uninitialized memory?), which is surprising.

Perhaps FlexAttention should either (possibly controlled by an option?):

- Raise an exception if the block mask does not match sequence length with the shape of the actual tensor being executed
- Or, automatically call `adjust` in such a scenario

Currently it appears to do neither, so my repro continues to unexpectedly read uninitialized memory.

**Comment by Chillee** [comment]
No, @yanboliang discussed it with us, and we agreed that this isn't the right approach.

Lemme just close this PR for now. @yanboliang feel free to re-open if you wanna reuse this PR.

**Comment by yanboliang** [comment]
Yea, I'm working on another more proper fix, would open anther PR soon.
